testplan/testing/environment/graph.py

```diff
@@ -0,0 +1,138 @@
+from copy import copy
+from itertools import product
+from typing import TYPE_CHECKING, Iterable, List
+
+from testplan.common.utils.graph import DirectedGraph
+
+if TYPE_CHECKING:
+    from testplan.testing.multitest.driver import Driver
+
+
+def _type_err(msg: str) -> TypeError:
+    return TypeError(
+        f"Wrong type used in Testplan Driver dependencies definition. {msg}"
+    )
+
+
+class DriverDepGraph(DirectedGraph[str, "Driver", bool]):
+    """
+    An always acyclic directed graph, also bookkeeping driver starting status.
+    """
+
+    _starting = set()
+
+    @classmethod
+    def from_directed_graph(cls, g: DirectedGraph) -> "DriverDepGraph":
+        if len(g.cycles()):
+            raise ValueError(
+                "Bad Testplan Driver dependency definition. "
+                f"Cyclic dependency detected among {g.cycles()[0]}."
+            )
+        g_ = copy(g)
+        return cls(g_.vertices, g_.edges, g_.indegrees, g_.outdegrees)
+
+    def mark_starting(self, driver: "Driver"):
+        self._starting.add(driver.uid())
+
+    def mark_started(self, driver: "Driver"):
+        self._starting.remove(driver.uid())
+        self.remove_vertex(driver.uid())
+
+    def drivers_to_start(self) -> List["Driver"]:
+        return [
+            self.vertices[d]
+            for d in self.zero_indegrees()
+            if d not in self._starting
+        ]
+
+    def drivers_starting(self) -> List["Driver"]:
+        return [self.vertices[d] for d in self._starting]
+
+    def all_drivers_started(self) -> bool:
+        return not len(self)
+
+    def purge_drivers_to_start(self):
+        """
+        A graph operation which purges everything in the graph except for the
+        drivers still in starting status.
+        """
+        self.vertices = {d: self.vertices[d] for d in self._starting}
+        self.edges = {d: dict() for d in self.vertices}
+        self.indegrees = {d: 0 for d in self.vertices}
+        self.outdegrees = {d: 0 for d in self.vertices}
+
+    def __copy__(self):
+        obj = self.__class__(
+            vertices=copy(self.vertices),
+            edges={src: copy(dst) for src, dst in self.edges.items()},
+            indegrees=copy(self.indegrees),
+            outdegrees=copy(self.outdegrees),
+        )
+        obj._starting = copy(self._starting)
```

Comment:
May have missed something. Why did we have to copy the `_starting` method as well?

Reply:
the current caller side doesn't need this, but i just want to make the copy magic method consistent with its name